api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,183 @@
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"reflect"
+	"time"
+
+	"go.opencensus.io/trace"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+// CloudEvent is the official JSON representation of a CloudEvent: https://github.com/cloudevents/spec/blob/master/serialization.md
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+type cloudEventIn struct {
+	CloudEvent
+
+	// Deadline string `json:"deadline"`
+	// Protocol CallRequestHTTP `json:"protocol"`
+}
+
+// cloudEventOut the expected response from the function container
+type cloudEventOut struct {
+	CloudEvent
+
+	// Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
+}
+
+// CloudEventProtocol converts stdin/stdout streams from HTTP into JSON format.
+type CloudEventProtocol struct {
+	// These are the container input streams, not the input from the request or the output for the response
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *CloudEventProtocol) IsStreamable() bool {
+	return true
+}
+
+func (h *CloudEventProtocol) writeJSONToContainer(ci CallInfo) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	_, err := io.Copy(buf, ci.Input())
+	if err != nil {
+		return err
+	}
+
+	in := cloudEventIn{
+		CloudEvent: CloudEvent{
```

Comment:
event type is required according to the spec https://github.com/cloudevents/spec/blob/master/spec.md#eventtype

Reply:
cloud event version is required according to the spec https://github.com/cloudevents/spec/blob/master/spec.md#cloudeventsversion

Reply:
source of the event is required as well according to the spec, this could be derived from the event type, so, if function was triggered with an HTTP request then we need to put a Host reader here.

Reply:
Fixed.